lib/customisation.nix

```
@@ -661,4 +662,171 @@ rec {
+      => "ab"
+  */
+  extendMkDerivation =
+    # Base build helper, the mkDerivation-like build helper to extend.
+    mkDerivationBase:
+    # An overlay of attribute set, like the one taken by <pkg>.overrideAttrs.
+    # This act as the implementation detail of the result build helper.
+    attrsOverlay:
+    mirrorFunctionArgs
+      # Make the __functionArgs looks like one from a build helper accepting plain attribute set.
+      (attrsOverlay { })
+      # Adds the fixed-point style support.
+      (fpargs: (mkDerivationBase fpargs).overrideAttrs attrsOverlay)
+    // {
+      # Passthru attributes attached to the result build helper.
+      inherit attrsOverlay;
+    };
+
+  /*
+    Like extendMkDerivation, but accept an argument set adapter instead of an
+    attribute overlay.
+
+    The argument set adapter returns the fixed attribute set to pass to the base
+    build helper, instead of a subet of attributes to update. This allows
+    removing arguments that we don't want to pass to the base build helper.
+
+    Type:
+      adaptMkDerivation ::
+        ((FixedPointArgs | AttrSet) -> a)
+        -> (AttrSet -> AttrSet -> AttrSet)
+        -> (FixedPointArgs | AttrSet) -> a
+
+      FixedPointArgs = AttrSet -> AttrSet
+      # a == Derivation when defining a build helper
+
+    Example:
+      mkLocalDerivation = lib.adaptMkDerivation pkgs.stdenv.mkDerivation (finalAttrs:
+        args@{ preferLocalBuild ? true, allowSubstitute ? false, specialArg ? (_: false), ... }:
+        removeAttrs args [ "specialArg" ] // { inherit preferLocalBuild allowSubstitute; })
+
+      mkLocalDerivation.__functionArgs
+      => { allowSubstitute = true; specialArg = true; preferLocalBuild = true; }
+
+      mkLocalDerivation { inherit (pkgs.hello) pname version src; specialArg = _: false; }
+      => «derivation /nix/store/xirl67m60ahg6jmzicx43a81g635g8z8-hello-2.12.1.drv»
+
+      mkLocalDerivation (finalAttrs: { inherit (pkgs.hello) pname version src; specialArg = _: false; })
+      => «derivation /nix/store/xirl67m60ahg6jmzicx43a81g635g8z8-hello-2.12.1.drv»
+
+      (mkLocalDerivation (finalAttrs: { inherit (pkgs.hello) pname version src; passthru = { foo = "a"; bar = "${finalAttrs.passthru.foo}b"; } })).bar
+      => "ab"
+  */
+
+  adaptMkDerivation =
+    # Base build helper, the mkDerivation-like build helper to extend.
+    mkDerivationBase:
+    # Argument set adapter, a function in the form `finalAttrs: args: { ... }`
+    # which turns the argument set of the result build helper into that for the
+    # base build helper.
+    # This acts as the implementation detail of the result build helper.
+    adaptArgs:
+    mirrorFunctionArgs
+      # Make the __functionArgs looks like one from a build helper accepting plain attribute set.
+      (adaptArgs { })
+      # Adds the fixed-point style support
+      (fpargs: mkDerivationBase (finalAttrs:
+        adaptArgs finalAttrs (lib.toFunction fpargs finalAttrs)
```

nit: `toFunction`  [is][1] in scope
```suggestion
        adaptArgs finalAttrs (toFunction fpargs finalAttrs)
```

[1]: https://github.com/NixOS/nixpkgs/pull/234651/files#diff-bfe3742899d9abd995e3b03ca21225dfc95b9d94d49898c45f2dfc435428c140R11